### DISCUSSION TITLE
1.7.0

### DISCUSSION THREAD

**Discussion by hjgraca** [aws-powertools/powertools-lambda-dotnet#476] [General]
## Summary

In this release we are please to announce the removal of the Common dependency and Nuget package 🎉 and introducing NuGet [Central Package Management](https://devblogs.microsoft.com/nuget/introducing-central-package-management/) (i.e. with Directory.Packages.props).

Utilities have a dependency on the AWS.Lambda.Powertools.Common project.
When the utility is compiled and packaged into a nuget package the AWS.Lambda.Powertools.Common is a dependency.
This behaviour is causing [Diamond dependency issues](https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/dependencies) in the project.
To address this we copy the required files to each utility and make them completely isolated. This allows the client to not have to worry about dependencies between utilities and AWS.Lambda.Powertools.Common.
There is no need to delete the AWS.Lambda.Powertools.Common project but link the files in other projects, this will make it more readable in the solution explorer and makes maintenance easier because it's all in a single project.

### TLDR

- Add `Directory.Build.props` and `Directory.Build.targets` [more info](https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-by-directory?view=vs-2022#directorybuildprops-and-directorybuildtargets)
	- These two files allows us to modify all csproj files in the solution or where the files are placed in the folder structure
	- Those files were added to the `src` folder. They modify all csproj in that folder where all utilities are. [Search scope](https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-by-directory?view=vs-2022#search-scope)
- Add Central Package Management [more info](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management)
	- Dependency management is a core feature of NuGet. Managing dependencies for a single project can be easy. Managing dependencies for multi-project solutions can prove to be difficult as they start to scale in size and complexity. In situations where you manage common dependencies for many different projects, you can leverage NuGet's central package management (CPM) features to do all of this from the ease of a single location.	
	- Add a file `Directory.Packages.props` to the src and test folders. This allows us to manage nuget package versions for all projects in a central location. I separated src and tests because tests have many dependencies and can cause confusion. [Search scope info](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management#central-package-management-rules)

#### Part 1 `Directory.Build.props`

- Use MSbuild to:
	- Manage all common properties across all projects
	- These properties can be overridden by the project if it wishes to do so
![image](https://github.com/aws-powertools/powertools-lambda-dotnet/assets/999396/754cceb6-d426-4dc1-b277-29e6f281b03e)

#### Part 2 `Directory.Build.targets` (**changes only happen in Release configuration**)

- Use MSbuild to:
	- Copy/Link `AWS.Lambda.Powertools.Common` *.cs files to the destination project
	- Remove depdendency of Common project by removing `<ProjectReference Remove="..\AWS.Lambda.Powertools.Common\AWS.Lambda.Powertools.Common.csproj" />`
	- Add PackageReference `<PackageReference Include="AspectInjector" />`
- This transformation only happens in **Release** configuration

![image](https://github.com/aws-powertools/powertools-lambda-dotnet/assets/999396/78488027-bebe-4e68-b03e-600712e5ef36)

Once in `Release` configuration or when running `dotnet build -c Release` the Common folder is added to the project

![image](https://github.com/aws-powertools/powertools-lambda-dotnet/assets/999396/610dc4df-6cc0-48b6-81b8-977f5d0fab84)

#### Part 3 Central Package Management

[Docs](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management)

Starting with [NuGet 6.2](https://learn.microsoft.com/en-us/nuget/release-notes/nuget-6.2), you can centrally manage your dependencies in your projects with the addition of a Directory.Packages.props file and an MSBuild property.

This means that the versions of the referenced packages will be defined in the central `Directory.Packages.props` file

File contents:
![image](https://github.com/aws-powertools/powertools-lambda-dotnet/assets/999396/289b5202-3740-4b87-a127-6b3484a5264c)

In the csproj files of the projects the referenced packages will not have a version

![image](https://github.com/aws-powertools/powertools-lambda-dotnet/assets/999396/c9b99114-dc32-4bc0-87df-1ab2e3968514)

If needed individual projects can [override the version](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management#overriding-package-versions)

#### Part 4 Nuget package contents

The resulting **Release** package will have no `AWS.Lambda.Powertools.Common` reference and will include `AspectInjector` nuget package. This will make the utility completely independent from `AWS.Lambda.Powertools.Common` Nuget package at runtime

**Old:**

![image](https://github.com/aws-powertools/powertools-lambda-dotnet/assets/999396/92efdced-f020-4584-b170-1c1b99ac5b98)

**New:**

![image](https://github.com/aws-powertools/powertools-lambda-dotnet/assets/999396/d00667b6-e86d-464f-bf09-e4ed7661f074)

Removed `AWS.Lambda.Powertools.Common` project reference from test projects.
This reference now comes from `Directory.Build.props`. This file was added to test folder.

Contents:
![carbon (1)](https://github.com/aws-powertools/powertools-lambda-dotnet/assets/999396/fa348431-1253-4205-8508-c27c2c5a441a)

## 🌟New features and non-breaking changes

* chore: Remove dependency of AWS.Lambda.Powertools.Common on all Utilities (#340) by @hjgraca

## 🔧 Maintenance

* chore: Disable dependabot.yml (#475) by @hjgraca
* chore: Update version release 1.7.0 (#464) by @hjgraca
* chore(deps): bump gitpython from 3.1.30 to 3.1.35 (#451) by @dependabot
* chore: update changelog with latest changes (#437) by @hjgraca
* chore: bump examples NuGet versions (#435) by @amirkaws

## This release was made possible by the following contributors:

@hjgraca and @amirkaws

<hr /><em>This discussion was created from the release <a href='https://github.com/aws-powertools/powertools-lambda-dotnet/releases/tag/1.7.0'>1.7.0</a>.</em>
